spec.md

```diff
@@ -19,15 +19,242 @@
+
+Example error response (invalid authorization):
+
+Example error response (missing or insufficient onboarding material provided):
+
+Example error response (invalid `id` provided):
+
+Example error response (a space with the specified `id` already exists):
+
+#### (Javascript API) Space .create()
+
+### Read Space operation
+
+* Requires appropriate authorization (root zcap invoked by the space's controller,
+  or a zcap granting permission to read a particular space)
+* Returns the details for the specified space `id`
+
+#### (HTTP API) GET `/spaces/{space_id}`
+
+Example request:
+
+```http
+GET /spaces/81246131-69a4-45ab-9bff-9c946b59cf2e HTTP/1.1
+Host: example.com
+Accept: application/json
+Authorization: ...
+```
+
+Example success response:
+
+```http
+HTTP/1.1 200 OK
+Content-type: application/json
+
+{
+  "id": "81246131-69a4-45ab-9bff-9c946b59cf2e",
+  "type": ["Space"],
+  "name": "Example space #1",
+  "controller": "did:key:z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW"
+}
+```
+
+Example error response (missing or insufficient authorization):
+
+Example error response (space id not found):
+
+#### (HTTP API) HEAD `/spaces/{space_id}`
+
+### List Spaces Operation
+
+* Requires appropriate authorization (root zcap invoked by the controller of one
+  or more spaces, or a zcap granting permission to read a one or more spaces)
+
+* Lists only the spaces the requester is authorized to see
+
+#### (HTTP API) GET `/spaces/`
+
+Example request:
+
+```http
+GET /spaces/ HTTP/1.1
+Host: example.com
+Accept: application/json
+Authorization: ...
+```
+
+Example success response (requester has read access to at least one space):
+
+```http
+HTTP/1.1 200 OK
+Content-type: application/json
+
+...
+```
+
+Example success response (requester does NOT have access to any spaces):
+
+```http
+HTTP/1.1 200 OK
+Content-type: application/json
+
+...
+```
+
+Example error response (missing authorization):
+
+### Update Space operation
+
+* Requires appropriate authorization (root zcap invoked by the space's controller,
+  or a zcap granting permission to write to a particular space)
+* Allows to update the following fields:
+  - `name`
+  - `controller`
+
+#### (HTTP API) PUT `/spaces/{space_id}`
+
+Note that this is a _full_ update (partial updates via http `PATCH` verb might
+be supported later). However, some fields may not be updated (like `id`) and so
+may be omitted from the request payload.
+
+Note that this operation is idempotent.
+
+Example request (updating the name of a space):
+
+```http
+PUT /spaces/81246131-69a4-45ab-9bff-9c946b59cf2e HTTP/1.1
+Host: example.com
+Content-type: application/json
+Accept: application/json
+Authorization: ...
+
+{
+  "id": "81246131-69a4-45ab-9bff-9c946b59cf2e",
+  "name": "Newly renamed space #1",
+  "controller": "did:key:z6MkpBMbMaRSv5nsgifRAwEKvHHoiKDMhiAHShTFNmkJNdVW"
+}
+```
+
+Example success response (returns the updated details of the space):
```

Comment:
Can this be a 204? I dont want to require the server to pay the bandwidth here. The client can opt into it by sending a GET

Reply:
Sure. It's a minor pedantic nicety (to return the new state after PUT), but you make a good point about saving bandwidth